Tests/Unit/Hooks/HookProviderTest.php

```diff
@@ -0,0 +1,248 @@
+    }
+
+    /**
+     * @test
+     */
+    public function hookImplementorImplementsHookHierachy()
+    {
+        $implementor = $this->createTestingHookImplementor();
+
+        self::assertInstanceOf(Hook::class, $implementor);
+        self::assertInstanceOf(TestingHookInterface::class, $implementor);
+    }
+
+    /**
+     * @test
+     */
+    public function hookImplementorTestHookMethodCanBeCalledAndReportsBeingCalled()
+    {
+        $implementor = $this->createTestingHookImplementor();
+
+        $implementor->testHookMethod();
+
+        self::assertSame(1, TestingHookImplementor::$wasCalled);
+    }
+
+    /*
+     * Tests concerning the TestingHookImplementor2
+     */
+
+    /**
+     * @test
+     */
+    public function hookImplementor2CanBeCreated()
+    {
+        self::assertInstanceOf(TestingHookImplementor2::class, $this->createTestingHookImplementor2());
+    }
+
+    /**
+     * @test
+     */
+    public function hookImplementor2ResetsCallCounterOnCreation()
+    {
+        TestingHookImplementor2::$wasCalled = 1;
+
+        $this->createTestingHookImplementor2();
+
+        self::assertSame(0, TestingHookImplementor2::$wasCalled);
+    }
+
+    /**
+     * @test
+     */
+    public function hookImplementor2ImplementsHookHierachy()
+    {
+        $implementor = $this->createTestingHookImplementor2();
+
+        self::assertInstanceOf(Hook::class, $implementor);
+        self::assertInstanceOf(TestingHookInterface::class, $implementor);
+    }
+
+    /**
+     * @test
+     */
+    public function hookImplementor2TestHookMethodCanBeCalledAndReportsBeingCalled()
+    {
+        $implementor = $this->createTestingHookImplementor2();
+
+        $implementor->testHookMethod();
+
+        self::assertSame(1, TestingHookImplementor2::$wasCalled);
+    }
+
+    /*
+     * Tests concerning the Hook object
+     */
+
+    /**
+     * @test
+     */
+    public function hookObjectForTestHookCanBeCreated()
+    {
+        self::assertInstanceOf(HookProvider::class, $this->createHookObject());
+    }
+
+    /**
+     * @test
+     * @doesNotPerformAssertions
+     */
+    public function hookObjectForTestHookWithNoHookImplemetorRegisteredSucceedsForValidMethod()
+    {
+        unset($GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['seminars'][TestingHookInterface::class]);
+        $hookObject = $this->createHookObject();
+
+        $hookObject->executeHook('testHookMethod');
+    }
+
+    /**
+     * @test
+     * @expectedException InvalidArgumentException
+     * @expectedExceptionCode 1573479911
+     */
+    public function hookObjectForTestHookWithNoHookImplemetorRegisteredFailsForEmptyMethod()
+    {
+        unset($GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['seminars'][TestingHookInterface::class]);
+        $hookObject = $this->createHookObject();
+
+        $hookObject->executeHook('');
+    }
+
+    /**
+     * @test
+     * @expectedException UnexpectedValueException
+     * @expectedExceptionCode 1573480302
```

Comment:
Generally, please use `$this->expectException`. The annotations are not recommended anymore. 

I don't think we need the checks for the exception code - I think that checking that an instance of a certain exception is thrown is sufficient.

Reply:
I'll continue checking the code, because

1. I easily find the exception I test for in the subject
2. I know I actually trigger that certain exception and nothing else along the path
3. I easily find the test for that particular exception if looking through the subject

Reply:
That's fair.